# Sparkify Project Workspace
This workspace contains a tiny subset (128MB) of the full dataset available (12GB). Feel free to use this workspace to build your project, or to explore a smaller subset with Spark before deploying your cluster on the cloud. Instructions for setting up your Spark cluster is included in the last lesson of the Extracurricular Spark Course content.

You can follow the steps below to guide your data analysis and model building portion of this project.

In [74]:
# import libraries
from pyspark.sql import SparkSession
import seaborn as sns
import matplotlib.pyplot as plt

In [3]:
# create a Spark session
from pyspark.sql.functions import udf
from pyspark.sql.types import StringType
from pyspark.sql.types import IntegerType
from pyspark.sql.functions import desc
from pyspark.sql.functions import asc
from pyspark.sql.functions import sum as Fsum

import datetime

from pyspark.sql import Window
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline

import re
from pyspark.sql import functions as sF
from pyspark.sql import types as sT

from functools import reduce

In [4]:
#Create Spark Session
spark = SparkSession.builder.appName("Sparkify").getOrCreate()

# Load and Clean Dataset
In this workspace, the mini-dataset file is `mini_sparkify_event_data.json`. Load and clean the dataset, checking for invalid or missing data - for example, records without userids or sessionids. 

In [5]:
#Read in data
data_link = 'mini_sparkify_event_data.json'
df = spark.read.json(data_link)

In [15]:
# Tabulate data 
df_P = df.toPandas()

In [17]:
df_P.head()

artist       auth firstName gender  itemInSession lastName  \
0    Martha Tilston  Logged In     Colin      M             50  Freeman   
1  Five Iron Frenzy  Logged In     Micah      M             79     Long   
2      Adam Lambert  Logged In     Colin      M             51  Freeman   
3            Enigma  Logged In     Micah      M             80     Long   
4         Daft Punk  Logged In     Colin      M             52  Freeman   

      length level                        location method      page  \
0  277.89016  paid                 Bakersfield, CA    PUT  NextSong   
1  236.09424  free  Boston-Cambridge-Newton, MA-NH    PUT  NextSong   
2  282.82730  paid                 Bakersfield, CA    PUT  NextSong   
3  262.71302  free  Boston-Cambridge-Newton, MA-NH    PUT  NextSong   
4  223.60771  paid                 Bakersfield, CA    PUT  NextSong   

   registration  sessionId                           song  status  \
0  1.538173e+12         29                      Rockpools     200   
1  1.538332e+12          8                         Canada     200   
2  1.538173e+12         29              Time For Miracles     200   
3  1.538332e+12          8    Knocking On Forbidden Doors     200   
4  1.538173e+12         29  Harder Better Faster Stronger     200   

              ts                                          userAgent userId  
0  1538352117000  Mozilla/5.0 (Windows NT 6.1; WOW64; rv:31.0) G...     30  
1  1538352180000  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...      9  
2  1538352394000  Mozilla/5.0 (Windows NT 6.1; WOW64; rv:31.0) G...     30  
3  1538352416000  "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebK...      9  
4  1538352676000  Mozilla/5.0 (Windows NT 6.1; WOW64; rv:31.0) G...     30

In [27]:
df_P.tail()

artist        auth firstName gender  itemInSession lastName  \
286495            None  Logged Out      None   None             41     None   
286496            None  Logged Out      None   None             42     None   
286497            None   Logged In    Emilia      F             43    House   
286498            None   Logged In    Emilia      F             44    House   
286499  Camera Obscura   Logged In    Emilia      F             45    House   

           length level                               location method  \
286495        NaN  paid                                   None    GET   
286496        NaN  paid                                   None    PUT   
286497        NaN  paid  New York-Newark-Jersey City, NY-NJ-PA    GET   
286498        NaN  paid  New York-Newark-Jersey City, NY-NJ-PA    GET   
286499  170.89261  paid  New York-Newark-Jersey City, NY-NJ-PA    PUT   

            page  registration  sessionId                 song  status  \
286495      Home           NaN        500                 None     200   
286496     Login           NaN        500                 None     307   
286497      Home  1.538337e+12        500                 None     200   
286498     About  1.538337e+12        500                 None     200   
286499  NextSong  1.538337e+12        500  The Sun On His Back     200   

                   ts                                          userAgent  \
286495  1543622240000                                               None   
286496  1543622241000                                               None   
286497  1543622248000  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
286498  1543622398000  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
286499  1543622411000  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   

        userId  
286495          
286496          
286497  300011  
286498  300011  
286499  300011

In [18]:
df_P.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286500 entries, 0 to 286499
Data columns (total 18 columns):
artist           228108 non-null object
auth             286500 non-null object
firstName        278154 non-null object
gender           278154 non-null object
itemInSession    286500 non-null int64
lastName         278154 non-null object
length           228108 non-null float64
level            286500 non-null object
location         278154 non-null object
method           286500 non-null object
page             286500 non-null object
registration     278154 non-null float64
sessionId        286500 non-null int64
song             228108 non-null object
status           286500 non-null int64
ts               286500 non-null int64
userAgent        278154 non-null object
userId           286500 non-null object
dtypes: float64(2), int64(4), object(12)
memory usage: 39.3+ MB


- Missing values for 9 columns / Same amount of missing values 
- Song and Artist are the same with 228108
- Datatypes may need to be changed 

In [28]:
df.printSchema()

root
 |-- artist: string (nullable = true)
 |-- auth: string (nullable = true)
 |-- firstName: string (nullable = true)
 |-- gender: string (nullable = true)
 |-- itemInSession: long (nullable = true)
 |-- lastName: string (nullable = true)
 |-- length: double (nullable = true)
 |-- level: string (nullable = true)
 |-- location: string (nullable = true)
 |-- method: string (nullable = true)
 |-- page: string (nullable = true)
 |-- registration: long (nullable = true)
 |-- sessionId: long (nullable = true)
 |-- song: string (nullable = true)
 |-- status: long (nullable = true)
 |-- ts: long (nullable = true)
 |-- userAgent: string (nullable = true)
 |-- userId: string (nullable = true)



In [26]:
# Check for duplicates 
df_P.duplicated().sum()

0

In [36]:
# Check for missing values 
df_P.isna().sum()

artist           58392
auth                 0
firstName         8346
gender            8346
itemInSession        0
lastName          8346
length           58392
level                0
location          8346
method               0
page                 0
registration      8346
sessionId            0
song             58392
status               0
ts                   0
userAgent         8346
userId               0
dtype: int64

In [53]:
# Explored nested columns 
unique_P = df_P['level'].unique(), df_P['method'].unique(), df_P['status'].unique(), df_P['auth'].unique(), df_P['gender'].unique(), df_P['page'].unique()

In [54]:
unique_P

(array(['paid', 'free'], dtype=object),
 array(['PUT', 'GET'], dtype=object),
 array([200, 307, 404]),
 array(['Logged In', 'Logged Out', 'Cancelled', 'Guest'], dtype=object),
 array(['M', 'F', None], dtype=object),
 array(['NextSong', 'Add to Playlist', 'Roll Advert', 'Thumbs Up',
        'Downgrade', 'Thumbs Down', 'Home', 'Logout', 'Help', 'Login',
        'Upgrade', 'Add Friend', 'About', 'Settings', 'Submit Upgrade',
        'Submit Downgrade', 'Error', 'Save Settings', 'Cancel',
        'Cancellation Confirmation', 'Register', 'Submit Registration'], dtype=object))

# Exploratory Data Analysis
When you're working with the full dataset, perform EDA by loading a small subset of the data and doing basic manipulations within Spark. In this workspace, you are already provided a small subset of data you can explore.

### Define Churn

Once you've done some preliminary analysis, create a column `Churn` to use as the label for your model. I suggest using the `Cancellation Confirmation` events to define your churn, which happen for both paid and free users. As a bonus task, you can also look into the `Downgrade` events.

### Explore Data
Once you've defined churn, perform some exploratory data analysis to observe the behavior for users who stayed vs users who churned. You can start by exploring aggregates on these two groups of users, observing how much of a specific action they experienced per a certain time unit or number of songs played.

# Feature Engineering
Once you've familiarized yourself with the data, build out the features you find promising to train your model on. To work with the full dataset, you can follow the following steps.
- Write a script to extract the necessary features from the smaller subset of data
- Ensure that your script is scalable, using the best practices discussed in Lesson 3
- Try your script on the full data set, debugging your script if necessary

If you are working in the classroom workspace, you can just extract features based on the small subset of data contained here. Be sure to transfer over this work to the larger dataset when you work on your Spark cluster.

# Modeling
Split the full dataset into train, test, and validation sets. Test out several of the machine learning methods you learned. Evaluate the accuracy of the various models, tuning parameters as necessary. Determine your winning model based on test accuracy and report results on the validation set. Since the churned users are a fairly small subset, I suggest using F1 score as the metric to optimize.

# Final Steps
Clean up your code, adding comments and renaming variables to make the code easier to read and maintain. Refer to the Spark Project Overview page and Data Scientist Capstone Project Rubric to make sure you are including all components of the capstone project and meet all expectations. Remember, this includes thorough documentation in a README file in a Github repository, as well as a web app or blog post.